apps/confidential/confidential_test.go

```diff
@@ -66,3 +75,105 @@ lQFhpxvPgPNBuxVRnsxoH/sLOA==
 		t.Fatalf("TestCertFromPEM: got nil key, want key != nil")
 	}
 }
+
+const token = "fake_token"
+
+var tokenScope = []string{"the_scope"}
+
+func fakeClient() (Client, error) {
+	cred, err := NewCredFromSecret("fake_secret")
+	if err != nil {
+		return Client{}, err
+	}
+	client, err := New("fake_client_id", cred, WithAuthority("https://fake_authority/fake"))
+	if err != nil {
+		return Client{}, err
+	}
+	client.base.Token.AccessTokens = &fake.AccessTokens{
+		AccessToken: accesstokens.TokenResponse{
+			AccessToken:   token,
+			ExpiresOn:     internalTime.DurationTime{T: time.Now().Add(1 * time.Hour)},
+			ExtExpiresOn:  internalTime.DurationTime{T: time.Now().Add(1 * time.Hour)},
+			GrantedScopes: accesstokens.Scopes{Slice: tokenScope},
+		},
+	}
+	client.base.Token.Authority = &fake.Authority{
+		InstanceResp: authority.InstanceDiscoveryResponse{
+			TenantDiscoveryEndpoint: "https://fake_authority/fake/discovery/endpoint",
+			Metadata: []authority.InstanceDiscoveryMetadata{
+				{
+					PreferredNetwork: "fake_authority",
+					PreferredCache:   "fake_cache",
+					Aliases: []string{
+						"fake_authority",
+						"fake_auth",
+						"fk_au",
+					},
+				},
+			},
+			AdditionalFields: map[string]interface{}{
+				"api-version": "2020-02-02",
+			},
+		},
+	}
+	client.base.Token.Resolver = &fake.ResolveEndpoints{
+		Endpoints: authority.NewEndpoints("https://fake_authority/fake/auth",
+			"https://fake_authority/fake/token", "https://fake_authority/fake/jwt", "fake_authority"),
+	}
+	client.base.Token.WSTrust = &fake.WSTrust{}
+	return client, nil
+}
+
+func TestAcquireTokenByCredential(t *testing.T) {
+	client, err := fakeClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = client.AcquireTokenSilent(context.Background(), tokenScope)
+	// first attempt should fail
+	if err == nil {
+		t.Fatal("unexpected nil error from AcquireTokenSilent")
+	}
+	tk, err := client.AcquireTokenByCredential(context.Background(), tokenScope)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if tk.AccessToken != token {
+		t.Fatalf("unexpected access token %s", tk.AccessToken)
+	}
+	// second attempt should return the cached token
+	tk, err = client.AcquireTokenSilent(context.Background(), tokenScope)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if tk.AccessToken != token {
+		t.Fatalf("unexpected access token %s", tk.AccessToken)
+	}
+}
+
+func TestAcquireTokenByAuthCode(t *testing.T) {
```

Comment:
This test will probably be a little different than s2s scenario. AcquireTokenByAuthCode in confidential client is the second step of Authorization code flow. 

In real use case scenario, it probably is a webapp which has done the first step of acquiring an auth code on its own.  And then passes that auth code to this function. That also makes me rethink that code cannot be an optional parameter here. We probably need to review this. Also this means that it can have a user present too.

Reply:
Yes agreed, the code should not be optional here.  It's not in the [C# impl](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/src/client/Microsoft.Identity.Client/IConfidentialClientApplication.cs#L44).  I will make it a required param.

Reply:
Fixed